### PR TITLE
Width of the window title bar with custom frame #4024

### LINKED ISSUE #3806: Window title bar border disappears
Subject: Window title bar border disappears                                           
                                                                                      
Dear emilk,                                                                           
                                                                                      
When I select a window in my program,                                                 
the color of the selected window's title bar changes,                                 
but the border of the title bar disappears.                                           
                                                                                      
Thank you, emilk.                                                                                
                                                                                      
egui version: egui v0.25.0

### LINKED ISSUE #4024: Width of the window title bar with custom frame
**Describe the bug**
While testing #4002 (thanks for the fast fix!) on 0.26.1, I found another issue, or at least I think it was not desired. When setting the inner margin and stroke, the title bar is wider than the window's content as can be seen below. I've put some ridiculous values to show it clearly, but it's visible even when the stroke is just a few pixels wide.

image

**To Reproduce**
1. Go to egui_demo_lib -> window_options.rs
2. Find creation of "demo_window_options" window
3. Make it as below:

```rust
use super::View as _;
let frame = egui::containers::Frame::none() // <-- new
	.fill(egui::Color32::from_rgb(180, 220, 160)) // <-- new
	.stroke(egui::Stroke::new(20.0, egui::Color32::from_rgb(40, 100, 30))) // <-- new
	.inner_margin(egui::Margin::symmetric(20.0, 20.0)); // <-- new
let mut window = egui::Window::new(title)
	.id(egui::Id::new("demo_window_options"))
	.frame(frame) // <-- new
	.resizable(resizable)
	.constrain(constrain)
	.collapsible(collapsible)
	.title_bar(title_bar)
	.scroll2(scroll2)
	.enabled(enabled);
```

**Expected behavior**
The title bar should respect margin/stroke settings and have the same width as the rest of the window.

### PR DESCRIPTION
**Pull request by rustbasic**
Closes #4024
Closes #3806

@emilk 

I think it would be more accurate if I could know `frame.stroke.width` from lines 1150 to 1156 of windows.rs.

You can change it to .
```
outer_range.min - frame.stroke.width;
outer_range.max - frame.stroke.width,
```

### CODE REVIEW

**Comment by rustbasic** [comment]
It's example for egui bug testing

```rust
// example for egui bug testing
use eframe::egui::*;

fn main() -> Result<(), eframe::Error> {
    let options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default().with_inner_size([800.0, 600.0]),
        ..Default::default()
    };

    eframe::run_simple_native("My egui App", options, move |ctx, _frame| {
        egui::CentralPanel::default().show(ctx, |ui| {
            ui.heading("My egui Application");

            let frame = egui::containers::Frame::none() // <-- new
                .fill(egui::Color32::from_rgb(180, 220, 160)) // <-- new
                .stroke(egui::Stroke::new(
                    20.0,
                    egui::Color32::from_rgb(40, 100, 30),
                )) // <-- new
                .inner_margin(egui::Margin::symmetric(20.0, 20.0)); // <-- new
            let _response_w = egui::Window::new("test")
                .id(egui::Id::new("demo_window_options1"))
                .frame(frame) // <-- new
                .resizable(true)
                .constrain(true)
                .collapsible(true)
                .title_bar(true)
                .scroll2([true, true])
                .enabled(true)
                .show(ui.ctx(), |ui| {
                    let (response, _painter) =
                        ui.allocate_painter(vec2(300.0, 150.0), Sense::click());
                    let clicked_pos = response.interact_pointer_pos();
                    if response.clicked() {
                        let text =
                            format!("clicked_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
                        ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                        ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                        ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                    }
                    if response.is_pointer_button_down_on() {
                        let text = format!("down_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
                        ui.label(text);
                    }
                });

            let frame = egui::containers::Frame::none() // <-- new
                .fill(egui::Color32::from_rgb(180, 220, 160)) // <-- new
                .stroke(egui::Stroke::new(1.0, egui::Color32::from_rgb(40, 100, 30))) // <-- new
                .inner_margin(egui::Margin::symmetric(20.0, 20.0)); // <-- new
            let _response_w = egui::Window::new("test")
                .id(egui::Id::new("demo_window_options2"))
                .frame(frame) // <-- new
                .resizable(true)
                .constrain(true)
                .collapsible(true)
                .title_bar(true)
                .scroll2([true, true])
                .enabled(true)
                .show(ui.ctx(), |ui| {
                    let (response, _painter) =
                        ui.allocate_painter(vec2(300.0, 150.0), Sense::click());
                    let clicked_pos = response.interact_pointer_pos();
                    if response.clicked() {
                        let text =
                            format!("clicked_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
                        ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                        ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                        ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                    }
                    if response.is_pointer_button_down_on() {
                        let text = format!("down_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
                        ui.label(text);
                    }
                });

            let (response, _painter) = ui.allocate_painter(vec2(300.0, 150.0), Sense::click());
            let clicked_pos = response.interact_pointer_pos();
            if response.clicked() {
                let text = format!("clicked_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
                ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
                ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
            }
            if response.is_pointer_button_down_on() {
                let text = format!("down_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
                ui.label(text);
            }
        });
    })
}
```

File: crates/egui/src/containers/window.rs [hunk 1]
```diff
@@ -514,10 +514,13 @@ impl<'open> Window<'open> {
             if let Some(title_bar) = title_bar {
                 if on_top && area_content_ui.visuals().window_highlight_topmost {
                     let rect = Rect::from_min_size(
-                        outer_rect.min,
```
Comment (YgorSouza) [inline crates/egui/src/containers/window.rs:517]: I think you can keep the original code to create the `Rect` and add `.shrink(frame_stroke.width)` at the end. Should make the intent more clear.

**Comment by YgorSouza** [comment]
Did you try with the default windows in the demo app as well? It seems a little off.

image

image

I think you want to shrink the rect by stroke/2.0, and you have to shrink the `Rounding` by that amount as well (maybe add a method for that)?

**Comment by rustbasic** [comment]
@YgorSouza @emilk 

YgorSouza, Thank you for the good advice. you are right.

That's all I'm going to do here.
I just wanted to save we the trouble of finding bugs.
Further work is requested from 'YgorSouza or emilk or someone else'.

**Comment by varphone** [comment]
> Closes #4024 Closes #3806
> 
> @emilk
> 
> I think it would be more accurate if I could know `frame.stroke.width` from lines 1150 to 1156 of windows.rs.
> 
> You can change it to .
> 
> ```
> outer_range.min - frame.stroke.width;
> outer_range.max - frame.stroke.width,
> ```



> 
> 
> It's example for egui bug testing
> 
> ```rust
> // example for egui bug testing
> use eframe::egui::*;
> 
> fn main() -> Result<(), eframe::Error> {
>     let options = eframe::NativeOptions {
>         viewport: egui::ViewportBuilder::default().with_inner_size([800.0, 600.0]),
>         ..Default::default()
>     };
> 
>     eframe::run_simple_native("My egui App", options, move |ctx, _frame| {
>         egui::CentralPanel::default().show(ctx, |ui| {
>             ui.heading("My egui Application");
> 
>             let frame = egui::containers::Frame::none() // <-- new
>                 .fill(egui::Color32::from_rgb(180, 220, 160)) // <-- new
>                 .stroke(egui::Stroke::new(
>                     20.0,
>                     egui::Color32::from_rgb(40, 100, 30),
>                 )) // <-- new
>                 .inner_margin(egui::Margin::symmetric(20.0, 20.0)); // <-- new
>             let _response_w = egui::Window::new("test")
>                 .id(egui::Id::new("demo_window_options1"))
>                 .frame(frame) // <-- new
>                 .resizable(true)
>                 .constrain(true)
>                 .collapsible(true)
>                 .title_bar(true)
>                 .scroll2([true, true])
>                 .enabled(true)
>                 .show(ui.ctx(), |ui| {
>                     let (response, _painter) =
>                         ui.allocate_painter(vec2(300.0, 150.0), Sense::click());
>                     let clicked_pos = response.interact_pointer_pos();
>                     if response.clicked() {
>                         let text =
>                             format!("clicked_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
>                         ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                         ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                         ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                     }
>                     if response.is_pointer_button_down_on() {
>                         let text = format!("down_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
>                         ui.label(text);
>                     }
>                 });
> 
>             let frame = egui::containers::Frame::none() // <-- new
>                 .fill(egui::Color32::from_rgb(180, 220, 160)) // <-- new
>                 .stroke(egui::Stroke::new(1.0, egui::Color32::from_rgb(40, 100, 30))) // <-- new
>                 .inner_margin(egui::Margin::symmetric(20.0, 20.0)); // <-- new
>             let _response_w = egui::Window::new("test")
>                 .id(egui::Id::new("demo_window_options2"))
>                 .frame(frame) // <-- new
>                 .resizable(true)
>                 .constrain(true)
>                 .collapsible(true)
>                 .title_bar(true)
>                 .scroll2([true, true])
>                 .enabled(true)
>                 .show(ui.ctx(), |ui| {
>                     let (response, _painter) =
>                         ui.allocate_painter(vec2(300.0, 150.0), Sense::click());
>                     let clicked_pos = response.interact_pointer_pos();
>                     if response.clicked() {
>                         let text =
>                             format!("clicked_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
>                         ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                         ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                         ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                     }
>                     if response.is_pointer_button_down_on() {
>                         let text = format!("down_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
>                         ui.label(text);
>                     }
>                 });
> 
>             let (response, _painter) = ui.allocate_painter(vec2(300.0, 150.0), Sense::click());
>             let clicked_pos = response.interact_pointer_pos();
>             if response.clicked() {
>                 let text = format!("clicked_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
>                 ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                 ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>                 ui.label(RichText::new(text.clone()).color(Color32::LIGHT_RED));
>             }
>             if response.is_pointer_button_down_on() {
>                 let text = format!("down_pos: {:?}", clicked_pos.unwrap_or(pos2(0.0, 0.0)));
>                 ui.label(text);
>             }
>         });
>     })
> }
> ```

Perhaps it would make more sense to fix this with a background fill.

* [rustbasic:patch9](https://github.com/rustbasic/egui/tree/patch9)

[patch-with-size.webm](https://github.com/emilk/egui/assets/1274171/532ae7b9-b4ea-4754-9295-4d2eafbff101)

* [varphone:hotfix/window-custom-frame](https://github.com/varphone/egui/commit/9226bb31a13476491c22692e0414fd48f1d38675)

[patch-with-fill-bg.webm](https://github.com/emilk/egui/assets/1274171/a9d6e4c5-c349-4daa-bdbb-530e2edfe264)

**Comment by rustbasic** [comment]
@varphone 

I respect you.
It would be a good idea for you to solve this issue and submit a Pull Request.

**Comment by rustbasic** [comment]
@varphone 

We need something for the `South East`.

and egui::Window border

**Comment by rustbasic** [comment]
@varphone

We need something for the South East.

and, for egui::Window outer border with Normal window (not custom window)

**Comment by rustbasic** [comment]
@varphone @emilk 

The fix was completed by applying `varphone:hotfix/window-custom-frame` advised by varphone.

**Pull request closed without merging**